### PR TITLE
[Windows] Move to `FlutterCompositor` for rendering

### PR DESCRIPTION
**Pull request by loic-sharma**
This migrates the Windows embedder to `FlutterCompositor` so that the engine renders off-screen to a framebuffer instead of directly onto the window's surface. This will allow us to support platform views and multiple views on Windows.

<details>
<summary>Tests...</summary>

* Verify OpenGL compositor's raster time isn't regressed and memory increase is reasonable
* Software compositor's raster time and memory isn't regressed

Test device configurations
* [x] Windows 11 (hardware acceleration enabled/disabled)
* [x] Windows Arm64 (hardware acceleration enabled/disabled)
* [x] Windows 7 (hardware acceleration enabled/disabled, DWM enabled/disabled)

</details>

Addresses https://github.com/flutter/flutter/issues/128904

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/windows/compositor_opengl.cc
Comment (loic-sharma) [inline shell/platform/windows/compositor_opengl.cc:25]: The `CompositorOpenGL` mirrors Linux's backing store provider: https://github.com/flutter/engine/blob/101396fd3b8216acea95f09fa7c3fc8e97b38ab8/shell/platform/linux/fl_backing_store_provider.cc#L19-L102

File: shell/platform/windows/BUILD.gn
Comment (loic-sharma) [inline shell/platform/windows/BUILD.gn:142]: The Windows embedder uses Impeller's `ProcTableGLES` and `DescriptionGLES` to resolve OpenGLES functions and determine the GL version and extensions. This was the approach recommended by Impeller folks: https://discord.com/channels/608014603317936148/1173336353187041380/1182110838849536040

File: shell/platform/windows/compositor_software.cc
Comment (loic-sharma) [reply shell/platform/windows/compositor_software.cc:17]: @cbracken Do you have any recommendations here?

FWIW, previously our bitmaps were allocated by Skia using `calloc`: https://github.com/google/skia/blob/4f2180695a74d09b8fb732274b1403e059a28460/src/core/SkMallocPixelRef.cpp#L47-L50

Comment (cbracken) [reply shell/platform/windows/compositor_software.cc:17]: A nice thing about calloc (even just from a debugging point of view) is that it gives you memory initialized to a known value (zero'ed). I'd be tempted to lean that way from both a debugging and security point of view.

**Review by yaakovschectman — APPROVED** [review]
LGTM as long as we're going with the C stdlib memory management

File: shell/platform/windows/compositor_software.cc
Comment (loic-sharma) [reply shell/platform/windows/compositor_software.cc]: I switched to `std::calloc`, but I'll leave this open for other folks to chime in.

File: shell/platform/windows/compositor_software.cc
Comment (yaakovschectman) [inline shell/platform/windows/compositor_software.cc:39]: This reminds me:
This wouldn't be in this PR, but how about iterating through however many backing store layers there are and blending them per-pixel-alpha so we can composite an arbitrary number of layers? I'm doing something similar on my local branch working with platform views. If that sounds good, I can start a PR once this lands to add that here.

File: shell/platform/windows/compositor_opengl.cc
Comment (cbracken) [inline shell/platform/windows/compositor_opengl.cc:136]: Consider refactoring from `HasExtension`/`GetGlVersion` (how we check for `GL_BGRA8_EXT` format) to something like `description->GetSupportedTextureFormat()` which would return either `GL_BGRA8_EXT` (the if/else if conditions) or `GL_RGBA8` the else condition.

Alternatively there's the middle ground of something like `description->SupportsBgra8Ext()` or similar.

File: shell/platform/windows/compositor_opengl.cc
Comment (cbracken) [inline shell/platform/windows/compositor_opengl.cc:142]: This is totally fine; you could also just initialise the field directly and `.reset()` it if `!gl_->IsValid()`. That has the very minor advantage of keeping the initialisation and error handling for failed initialisation all in one place and avoids an extra local/move, but I don't feel at all strongly about it.

File: shell/platform/windows/compositor_opengl.h
Comment (cbracken) [inline shell/platform/windows/compositor_opengl.h:50]: Consider initialising to a default junk value like `0`.

File: shell/platform/windows/compositor_opengl_unittests.cc
Comment (cbracken) [inline shell/platform/windows/compositor_opengl_unittests.cc:29]: Here and below: `static_cast<unsigned char*>("3.0");

File: shell/platform/windows/compositor_opengl_unittests.cc
Comment (cbracken) [inline shell/platform/windows/compositor_opengl_unittests.cc:45]: Guessing this doesn't support `std::string_view` which would get you nice operator== semantics?

File: shell/platform/windows/compositor_opengl_unittests.cc
Comment (cbracken) [inline shell/platform/windows/compositor_opengl_unittests.cc:97]: As the reader, my first question would be "now that we're no longer the owner of `surface_manager`, is `EngineModifier` definitely freeing it? The answer is of course yes, but it may be worth making this more obvious at the call site. I can think of two ways to do that:

1. Add an `EngineModifier::SetSurfaceManager(std::unique_ptr<AngleSurfaceManager>)` override and `std::move(surface_manager)` here.
2. Add a comment `// Hand ownership of surface_manager to EngineModifier.`

File: shell/platform/windows/compositor_opengl_unittests.cc
Comment (cbracken) [inline shell/platform/windows/compositor_opengl_unittests.cc:107]: Question -- why use the local? This assignment is inevitable since we're using `EXPECT_CALL`, so the test will continue running even if either of those checks fails.

File: shell/platform/windows/compositor_software.cc
Comment (cbracken) [inline shell/platform/windows/compositor_software.cc:27]: Where is the wiring to call back into `CompositorSoftware::CollectBackingStore`? It might be worth adding a comment here explaining that.

File: shell/platform/windows/flutter_windows_engine.cc
Comment (cbracken) [inline shell/platform/windows/flutter_windows_engine.cc:404]: Ah I see - alright this is where to point to in the comment above :)

**Review by cbracken — APPROVED** [review]
Overall looks great -- just a few nits!

File: shell/platform/windows/compositor_opengl.cc
Comment (loic-sharma) [reply shell/platform/windows/compositor_opengl.cc:142]: Updated!

File: shell/platform/windows/compositor_opengl.h
Comment (loic-sharma) [reply shell/platform/windows/compositor_opengl.h]: Updated!

File: shell/platform/windows/compositor_opengl_unittests.cc
Comment (loic-sharma) [reply shell/platform/windows/compositor_opengl_unittests.cc]: One day I'll remember C++ style casts, but apparently that day is not today 😂

Fixed!

File: shell/platform/windows/compositor_opengl_unittests.cc
Comment (loic-sharma) [reply shell/platform/windows/compositor_opengl_unittests.cc:45]: Right, this mirrors `eglResolveFunction`. I updated this to allocate convert the `name` into a `std::string` to get those nice `operator==` semantics :)

File: shell/platform/windows/compositor_software.cc
Comment (loic-sharma) [reply shell/platform/windows/compositor_software.cc]: Added comments!

File: shell/platform/windows/flutter_windows_engine.cc
Comment (loic-sharma) [reply shell/platform/windows/flutter_windows_engine.cc:404]: Added comment

File: shell/platform/windows/compositor_opengl_unittests.cc
Comment (loic-sharma) [reply shell/platform/windows/compositor_opengl_unittests.cc:109]: I'm not sure I understand this feedback. These `EXPECT_CALL`s are necessary mocks for `FlutterWindowsView`'s constructor and need to be registered before we create the `FlutterWindowsView`. Please let me know if I misunderstood your question.

File: shell/platform/windows/compositor_software.cc
Comment (loic-sharma) [reply shell/platform/windows/compositor_software.cc:42]: That sounds reasonable but before that PR we'd first want a design doc shared and reviewed by everyone :)

File: shell/platform/windows/compositor_opengl_unittests.cc
Comment (loic-sharma) [reply shell/platform/windows/compositor_opengl_unittests.cc:99]: This is an excellent idea. I've opened https://github.com/flutter/engine/pull/48953 to clean this up. Once that lands I'll rebase and update this pull request.

File: shell/platform/windows/BUILD.gn
Comment (jonahwilliams) [reply shell/platform/windows/BUILD.gn:142]: why does impeller require the context to be current - do we combine creation of the proc table with looking up gl state? Could we split that functionality out of proc table so you could use it exclusively?

**Review by jonahwilliams — APPROVED** [review]
Usage looks fine to me. But if we can make an adjustment to proc table to make it more useful we should do that.

File: shell/platform/windows/compositor_opengl.cc
Comment (dkwingsmt) [inline shell/platform/windows/compositor_opengl.cc:82]: Maybe add a TODO here to explain that composition is not yet supported? As well as the assertion for the layer below.

File: shell/platform/windows/BUILD.gn
Comment (loic-sharma) [reply shell/platform/windows/BUILD.gn:142]: > do we combine creation of the proc table with looking up gl state? 

Yes. The proc table uses the `DescriptionGLES` to determine capabilities and [examine extensions' functions](https://github.com/flutter/engine/blob/ae61286e4dec66562590a2db2a421bc38e6492d2/impeller/renderer/backend/gles/proc_table_gles.cc#L119-L129), and the `DescriptionGLES` constructor fails if the context isn't current.

> Could we split that functionality out of proc table so you could use it exclusively?

I think so. We could have a "raw" proc table that can be initialized without a GL context, does not have the description/capabilities, and includes potentially unsupported extension functions. The "safe" proc table would load the description/capabilities and remove unsupported functions. This might require a little experimentation to get the API right.

File: shell/platform/windows/compositor_opengl.cc
Comment (loic-sharma) [reply shell/platform/windows/compositor_opengl.cc:84]: Added TODO!

File: shell/platform/windows/compositor_opengl_unittests.cc
Comment (loic-sharma) [reply shell/platform/windows/compositor_opengl_unittests.cc]: Fixed!

File: shell/platform/windows/compositor_opengl_unittests.cc
Comment (cbracken) [reply shell/platform/windows/compositor_opengl_unittests.cc:109]: Ah sorry - my bad. I missed that you're using a subtype here so you'd need narrowing casts if you made the assignment to `view_` directly without the local.

File: shell/platform/windows/compositor_opengl_unittests.cc
Comment (cbracken) [reply shell/platform/windows/compositor_opengl_unittests.cc:45]: You could even use std::string_view and get the operator== emantics roughly for free without an additional allocation/copy.

File: shell/platform/windows/compositor_opengl.cc
Comment (loic-sharma) [reply shell/platform/windows/compositor_opengl.cc]: I refactored this into a new function `GetSupportedTextureFormat(...)`, but I didn't put this on Impeller's `DescriptionGLES` as I'm not confident how to generalize this logic for all platforms Impeller runs on. For example, [Skia' logic for this is quite complex](https://github.com/google/skia/blob/4738ed711e03212aceec3cd502a4adb545f38e63/src/gpu/ganesh/gl/GrGLCaps.cpp#L1963-L2116).

Comment (cbracken) [reply shell/platform/windows/compositor_opengl.cc]: Nice. Yeah if this API is much more general, then a local static function sounds like the way to go!

**Comment by loic-sharma** [comment]
Reverting. The framework tree became red on `Windows run_debug_test_windows` and `Windows_arm64 run_debug_test_windows` once this commit rolled to the framework.

**Comment by wanjm** [comment]
why it is reverted?

**Pull request merged**
